### PR TITLE
[Object] Refactored Path-Formatter to use DI (incl. BC Layer)

### PR DESCRIPTION
**Pull request by dpfaffenbauer**
This PR refactors the implementation for PathFormatter classes by using a Interface and DI.

### CODE REVIEW

**Comment by dpfaffenbauer** [comment]
TO be honest, I am not happy with using the DI container directly in the Resolver class, It would be better to have a registry with all Formatters, register a new one with a DI Tag and display a dropdown list in the UI. But thats to much of a BC, right?

**Comment by weisswurstkanone** [comment]
This should be mentioned in the upgrade notes as well (as a BC break warning for pimcore 6)
doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md

**Comment by dpfaffenbauer** [comment]
will do end of october/start of november.

**Comment by dpfaffenbauer** [comment]
Just wondering what the next release will be. Will there be a Pimcore 5.6 or is the next release gonna be 6.0? Since this should be either come within a minor release or major release. If the next release is gonna be 6.0, there is no need for a BC layer right?

**Comment by dpfaffenbauer** [comment]
ping @weisswurstkanone

**Comment by dpfaffenbauer** [comment]
just added a note to the upgrade docs

**Comment by dpfaffenbauer** [comment]
ping @brusch will this ever get merged or can we just close it?

**Comment by brusch** [comment]
@dpfaffenbauer ok for me, can you resolve the conflicts? 
Thanks!

**Comment by brusch** [comment]
Ping 🏓

**Comment by dpfaffenbauer** [comment]
this Pr was open for months without any comment. Now you have to give me some time to change it.

**Comment by brusch** [comment]
No problem at all, I was just wondering if you got my update 😉

**Comment by dpfaffenbauer** [comment]
rebased onto master now

**Comment by dvesh3** [comment]
@dpfaffenbauer can you please check there are still conflicts? thanks!

**Comment by dpfaffenbauer** [comment]
done

File: bundles/AdminBundle/Controller/Admin/ElementController.php [hunk 1]
```diff
@@ -774,4 +745,77 @@ public function analyzePermissionsAction(Request $request)
...
+            if ($context['subContainerType'] == 'localizedfield') {
+                $fdLocalizedFields = $fdCollection->getFieldDefinition('localizedfields');
+                $fd = $fdLocalizedFields->getFieldDefinition($fieldname);
+            } else {
+                $fd = $fdCollection->getFieldDefinition($fieldname);
+            }
+        }
+
+
+        if ($fd instanceof DataObject\ClassDefinition\PathFormatterAwareInterface) {
```
Comment (dvesh3) [inline bundles/AdminBundle/Controller/Admin/ElementController.php]: @dpfaffenbauer can you please explain..how this suppose to work? afaik there are no field definitions implementing this interface.

File: doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md [hunk 2]
```diff
@@ -1,6 +1,9 @@
 # Upgrade Notes for Upgrades within Pimcore 5
 
 ## Version 5.8.0
+- Add PathFormatter interface, static path formatters are now deprecated and will be removed with Version 6.0. You can read
+ more about that [here](../../../05_Objects/01_Object_Classes/05_Class_Settings/14_Path_Formatter.md).
+- Removed method \Pimcore\Model\DataObject\ClassDefinition\Data::setFieldtype($fieldtype)
```
Comment (dvesh3) [inline doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md]: please remove duplicate docs :)

Comment (dpfaffenbauer) [reply doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md]: I don't see any duplicate here?

File: bundles/AdminBundle/Controller/Admin/ElementController.php (hunk 1 above)
Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/ElementController.php]: guess it got lost during the rebase, I re-added it

File: doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md (hunk 2 above)
Comment (dvesh3) [reply doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md]: It's about "setFieldtype" notice already mentioned in 5.6.0

File: bundles/AdminBundle/Controller/Admin/ElementController.php (hunk 1 above)
Comment (dvesh3) [reply bundles/AdminBundle/Controller/Admin/ElementController.php]: please also fix the exception. thanks! `Pimcore\Model\DataObject\ClassDefinition\Data\Relations\AbstractRelations::getPathFormatterClass() must be compatible with Pimcore\Model\DataObject\ClassDefinition\PathFormatterAwareInterface::getPathFormatterClass(): ?string`

Comment (dpfaffenbauer) [reply bundles/AdminBundle/Controller/Admin/ElementController.php]: done

File: doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md (hunk 2 above)
Comment (dpfaffenbauer) [reply doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/01_Within_V5/README.md]: done

File: models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        if ($formatterClass) {
+            $formatter = null;
+
+            if (isset(self::$formatterCache[$formatterClass])) {
+                return self::$formatterCache[$formatterClass];
+            }
+            if (substr($formatterClass, 0, 1) == '@') {
+                $serviceName = substr($formatterClass, 1);
+                try {
+                    $formatter = \Pimcore::getKernel()->getContainer()->get($serviceName);
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php]: formatter not resolved when using DI in Formatter Class config

Comment (dpfaffenbauer) [reply models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php]: why not? whats your test scenario?

Comment (dvesh3) [reply models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php]: Sorry, there was some issue with defining service at my end. maybe good idea is to add simple example for service configuration when using DI method.

Comment (dpfaffenbauer) [reply models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php]: it is and was always with `@SERVICENAME`...

**Comment by dvesh3** [comment]
@dpfaffenbauer It is not working when you 'Asnyc Load' along with PathFormatter. can you please check?

image

**Comment by dvesh3** [comment]
Hello @dpfaffenbauer any update?

**Comment by dpfaffenbauer** [comment]
not yet, will check this week

**Comment by dpfaffenbauer** [comment]
@diveshp3 just checked on why it doesn't work with async load in admin, and its not this PR's fault. It returns the right data, but doesn't display it:

Bildschirmfoto 2019-04-24 um 15 45 32

**Comment by dvesh3** [comment]
I know it is not your PR's fault. maybe regression of #4122 but we can't merge it without a fix!

**Comment by dpfaffenbauer** [comment]
So I have to fix it even though its not introduced by this PR?

File: models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php (hunk 3 above)
Comment (dvesh3) [reply models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php]: **Doc updates:**
I mean provide service example something like this https://github.com/pimcore/pimcore/blob/master/doc/Development_Documentation/05_Objects/01_Object_Classes/01_Data_Types/08_Dynamic_Select_Types.md 

Also update preferred way to use path formatter with DI and add deprecation message for old one

**Class definition:**
change lable to
image

Comment (dpfaffenbauer) [reply models/DataObject/ClassDefinition/Helper/PathFormatterResolver.php]: Deprecation Message: https://github.com/pimcore/pimcore/pull/3399/files#diff-385d3cb9b74d2e9378fedd615fd53648R797

Example: https://github.com/pimcore/pimcore/pull/3399/files#diff-9d97d6e06d55f5e9f21a271e3e61d282R17

label: https://github.com/pimcore/pimcore/pull/3399/files#diff-8b61221e64319cf4539f459c80217173R513

**Comment by dvesh3** [comment]
I just verified that `get-nice-path` response is not correct when using your PR so please have a look at changes done in ElementController.php from 6220433

correct response:
image

**Comment by dpfaffenbauer** [comment]
found it and fixed it

**Pull request merged**